### ISSUE TITLE
Add support custom oauth token provider

### ISSUE THREAD

**Issue by kdhfred** [karafka/karafka#1767]
## Is your feature request related to a problem? Please describe.

TL; DR
I can't implement a custom token provider in Karafka because there is no functionality around SetOAuthBearerToken, SetOAuthBearerTokenFailure.

----
Currently, I want to use MSK IAM.
AWS supports OAUTHBEARER as an MSK IAM authentication method.

AWS provides token providers for Python, Go, Js, and Net, and you can use those tokens to use MSK IAM.

ref: https://github.com/aws?q=signer&type=all&language=&sort=

I thought I could use `oauthbearer_refresh_token_cb` to inject the token, but I got the following error.

```
/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rdkafka-0.14.0/lib/rdkafka/config.rb:242:in `block (2 levels) in native_config': Property "oauthbearer_token_refresh_cb" must be set through dedicated .._set_..() function (Rdkafka::Config::ConfigError)
	from /Users/fred/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rdkafka-0.14.0/lib/rdkafka/config.rb:232:in `each'
	from /Users/fred/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rdkafka-0.14.0/lib/rdkafka/config.rb:232:in `block in native_config'
	from <internal:kernel>:90:in `tap'
	from /Users/fred/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rdkafka-0.14.0/lib/rdkafka/config.rb:230:in `native_config'
	from /Users/fred/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/rdkafka-0.14.0/lib/rdkafka/config.rb:151:in `consumer'
	from main.rb:26:in `<main>'
```

When I checked, there didn't seem to be any bindings related to the oauth refresh callback.
(If I'm wrong, please point it out).

I implemented the consumer with confluent-kafka-go, the Go binding of librdkafka, and it worked fine.

I am attaching the example code using confluent kafka go.

Example: https://github.com/kdhfred/aws-msk-iam-example/blob/main/go/main.go

## Describe the solution you'd like

Please binding SetOAuthBearerToken, SetOAuthBearerTokenFailure and return OAuthBearerTokenRefresh error to make custom token provider.

**Comment by mensfeld** [comment]
Also @kdhfred please ping me if you decide to work on it as I do not want to double efforts.

**Comment by mensfeld** [comment]
> But I'm not familiar with Ruby so I don't know how much time it will take, I'm thinking maybe 5 days roughly. How long do you think this will take?

It should not take longer than a day to add to karafka-rdkafka. One or two days to add proper events, hooks and patches to karafka. In the end specs. So I guess around 5 days yeah.

If you are not in a rush, I think it should be done by me on Q1 2024.

**Comment by kdhfred** [comment]
Oh I see. I don't need to rush right now. but I'll let you know whether I do or do not until this month.

**Comment by kdhfred** [comment]
@mensfeld sorry for late reply.
I decide to not use MSK/IAM for ruby client at this time to move forward. Just let me know me if you need a help or user testing about this issue.

**Comment by jlambert121** [comment]
I was just looking for this feature here, essentially identical to the OP and would be super useful for us!  I may be able to look into this in next month or so, but certainly can help test if someone else takes a look at it first.

**Comment by mensfeld** [comment]
As I already said:

> It is, however, not a priority for me. If you rely on this, consider supporting my work via the https://karafka.io/#become-pro so I can prioritize this for you, or feel free to PR.

I can provide this feature end-to-end within 2 weeks window if I get a declaration that someone that needs this is willing to support my work by getting the Pro offering.

Otherwise this will have to wait and 100% will be available but ETA is not defined. Probably end of Q1, maybe Q2. It's hard for me to say as there are other (OSS and Pro) things higher on my list of priorities.

If there is a PR for this callback in rdkafka-ruby and appropriate handlers for Karafka (aside from execution on start that is already done and I added it exactly for this reason) I am happy to review, accept and release.

**Comment by mensfeld** [comment]
sidenote for this feature: the same oauth token management is needed in waterdrop and it should be transient for the end user. That is, auto-configured or karafka config based waterdrop should auto-inherit the refresh flow.

**Comment by bruce-szalwinski-he** [comment]
> It should not take longer than a day to add to karafka-rdkafka. One or two days to add proper events, hooks and patches to karafka. In the end specs. So I guess around 5 days yeah.

I see both karafka-rdkafka and rdkafka-ruby repos under the `karafka` org.  The former is forked from the latter.  They both publish to the same gem, `rdkafka-ruby`.  I think `karafka-rdkafka` publishes lower version numbers then `rdkafka-ruby`.  Is the idea that `karafka-rdkafka` is pre release version of `rdkafka-ruby`?

**Comment by mensfeld** [comment]
>  Is the idea that karafka-rdkafka is pre release version of rdkafka-ruby?

No. I am slowly working on their full merger. It just takes time. Whatever goes to rdkafka-ruby is synced to karafka-rdkafka and whatever goes to karafka-rdkafka already has or will be synced. Once they are are the same (in terms of code) karafka-rdkafka will be retired and rdkafka-ruby will remain.

This distinction is a legacy thing that I want to solve this year.

**Comment by mensfeld** [comment]
@bruce-szalwinski-he implemented the rdkafka-ruby layer. I am now moving to implement the waterdrop and karafka components...

**Comment by mensfeld** [comment]
@bruce-szalwinski-he do you mind looking and testing the producer: https://github.com/karafka/waterdrop/pull/467 :pray:

**Comment by mensfeld** [comment]
> It should not take longer than a day to add to karafka-rdkafka


Oh how wrong was I :laughing:  nonetheless it is done. PR is pending review. Waterdrop has its done tested and released in beta

I will close this PR once karafka PR is merged.

**Comment by dnnyjns** [comment]
Perfect timing. My company is picking up Karafka and we need to use MSK IAM. Would it make sense to add example usage to the docs? I suppose this is a good start https://github.com/karafka/waterdrop/pull/467#issuecomment-2018850102

**Comment by mensfeld** [comment]
@dnnyjns RC releases for the whole ecosystem should be live today.

**Comment by jjatinggoyal** [comment]
@mensfeld karafka docs still mention that AWS IAM is not supported.
https://github.com/karafka/karafka/wiki/FAQ#can-i-use-aws-msk-serverless-with-iam-authentication

Can this be corrected? And examples usage added for IAM auth?

**Comment by jjatinggoyal** [comment]
@dnnyjns have you started using karafka with MSK IAM auth? If yes, could you share a more concrete usage example?

**Comment by mensfeld** [comment]
@jjatinggoyal this statement is still factually correct. Karafka does not support AWS IAM in any way. It supports custom oauth token providers https://github.com/karafka/karafka/wiki/Deployment#custom-oauth-token-providers

I will add a note that custom oauth token provider fetch for aws can be done but this is NOT out of the box solution and may require third party gems.

**Comment by jjatinggoyal** [comment]
Thanks @mensfeld , that'd be great.

**Comment by bruce-szalwinski-he** [comment]
Now that we've worked out how to auth, I'll update the example in https://github.com/bruce-szalwinski-he/aws-msk-iam-sasl-signer-ruby to reflect the API.  

I had an earlier conversation with a maintainer of ruby sdk at AWS.  Now that the sasl signer is complete, I'll ask him about the process for getting it adopted.

**Comment by mensfeld** [comment]
@jjatinggoyal I updated both FAQ and deployment docs.

@bruce-szalwinski-he once you are done with this I am happy to provide appropriate docs for it ofc :) (or happy to accept a PR :) )
